packages/sync-service/lib/electric/postgres/configuration.ex

```diff
@@ -7,40 +7,101 @@ defmodule Electric.Postgres.Configuration do
+  @type filters() :: %{Electric.oid_relation() => RelationFilter.t()}
 
   @pg_15 150_000
 
   @doc """
-  Ensure that all tables are configured for replication.
+  Configure the publication to include all relevant tables, also setting table identity to `FULL`
+  if necessary.
 
-  Table is considered configured for replication when it's `REPLICA IDENTITY` is set to `FULL`
-  and it's added to the specified publication.
+  Any tables that were previously configured but were either renamed or dropped will *not* be automatically
+  re-added to the publication.
+
+  `previous_relations` argument is used to figure out renamed/dropped table based on table OIDs.
 
   Important: this function should not be ran in a transaction, because it starts multiple
   internal transactions that are sometimes expected to fail.
 
   Raises if it fails to configure all the tables in the expected way.
   """
-  @spec configure_tables_for_replication!(
+  @spec configure_publication!(
           Postgrex.conn(),
+          [Electric.oid_relation()],
           filters(),
-          String.t(),
-          float()
-        ) ::
-          {:ok, [:ok]}
-  def configure_tables_for_replication!(pool, relation_filters, pg_version, publication_name) do
-    Postgrex.transaction(
-      pool,
-      fn conn ->
-        configure_tables_for_replication_internal!(
-          conn,
-          relation_filters,
-          pg_version,
-          publication_name
-        )
+          non_neg_integer(),
+          String.t()
+        ) :: [Electric.oid_relation()]
+  def configure_publication!(
+        conn,
+        previous_relations,
+        new_filters,
+        pg_version,
+        publication_name
+      ) do
+    Postgrex.transaction(conn, fn conn ->
+      # "New filters" were configured using a schema read in a different transaction (if at all, might have been from cache)
+      # so we need to check if any of the relations were dropped/renamed since then
+      changed_relations =
+        (previous_relations ++ Map.keys(new_filters))
+        |> Enum.uniq()
+        |> list_missing_relations(conn)
+
+      used_filters = Map.drop(new_filters, changed_relations)
+
+      if pg_version < @pg_15 do
+        alter_pub_set_whole_tables!(conn, publication_name, used_filters)
+      else
+        alter_pub_set_filtered_tables!(conn, publication_name, used_filters)
       end
-    )
+
+      # `ALTER TABLE` should be after the publication altering, because it takes out an exclusive lock over this table,
+      # but the publication altering takes out a shared lock on all mentioned tables, so a concurrent transaction will
+      # deadlock if the order is reversed.
+      set_replica_identity!(conn, used_filters)
+
+      changed_relations
+    end)
+    |> case do
+      {:ok, missing_relations} ->
+        missing_relations
+
+      {:error, reason} ->
+        raise reason
+    end
+  end
+
+  defp list_missing_relations(previous_relations, conn) do
```

Comment:
```suggestion
  defp list_missing_relations(relations, conn) do
```
nit: technically this helper doesn't know anything about previous or next relations - just whether the relations are present/the same

also technically this would be better if called `list_changed_relations`?